components/provisioner/internal/hydroform/service.go

```diff
@@ -0,0 +1,238 @@
+	credentialsFileName, err := c.saveCredentialsToFile(secretName)
+	if err != nil {
+		return ClusterInfo{}, err
+	}
+	defer removeFile(credentialsFileName)
+
+	log.Info("Preparing config for runtime provisioning")
+	cluster, provider, err := c.prepareConfig(runtimeConfig, credentialsFileName)
+	if err != nil {
+		return ClusterInfo{}, errors.Wrap(err, "Config preparation failed")
+	}
+
+	log.Infof("Config prepared - cluster: %+v, provider: %+v. Starting cluster provisioning", cluster, provider)
+
+	cluster, err = hf.Provision(cluster, provider)
+	if err != nil {
+		return ClusterInfo{}, errors.Wrap(err, "Cluster provisioning failed")
+	}
+
+	status, err := hf.Status(cluster, provider)
+	if err != nil {
+		return ClusterInfo{}, err
+	}
+
+	log.Info("Retrieving kubeconfig")
+
+	kubeconfig, err := hf.Credentials(cluster, provider)
+	if err != nil {
+		return ClusterInfo{}, errors.Wrap(err, "Failed to get kubeconfig")
+	}
+
+	log.Info("Retrieving cluster state")
+
+	internalState, err := stateToJson(cluster.ClusterInfo.InternalState)
+
+	if err != nil {
+		return ClusterInfo{}, errors.Wrap(err, "Failed to retrieve cluster state")
+	}
+
+	log.Infof("Cluster state: %+v", internalState)
+
+	return ClusterInfo{
+		ClusterStatus: status.Phase,
+		State:         internalState,
+		KubeConfig:    string(kubeconfig),
+	}, nil
+}
+
+func (c client) DeprovisionCluster(runtimeConfig model.RuntimeConfig, secretName string, terraformState string) error {
+	credentialsFileName, err := c.saveCredentialsToFile(secretName)
+	if err != nil {
+		return err
+	}
+	defer removeFile(credentialsFileName)
+
+	log.Info("Preparing config for runtime deprovisioning")
+	cluster, provider, err := c.prepareConfig(runtimeConfig, credentialsFileName)
+
+	if err != nil {
+		return errors.Wrap(err, "Config preparation failed")
+	}
+
+	state, err := jsonToState(terraformState)
+
+	if err != nil {
+		return errors.Wrap(err, "Config preparation failed")
+	}
+
+	cluster.ClusterInfo = &types.ClusterInfo{InternalState: state}
+
+	log.Infof("Config prepared - cluster: %+v, provider: %+v. Starting cluster deprovisioning", cluster, provider)
+	return hf.Deprovision(cluster, provider)
+}
+
+func jsonToState(state string) (*types.InternalState, error) {
+	var terraformState types.InternalState
+
+	err := json.Unmarshal([]byte(state), &terraformState)
+
+	if err != nil {
+		return &types.InternalState{}, err
+	}
+
+	return &terraformState, nil
+}
+
+func stateToJson(state *types.InternalState) (string, error) {
+	bytes, err := json.Marshal(state)
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+func (c client) saveCredentialsToFile(secretName string) (string, error) {
+	secret, err := c.secrets.Get(secretName, meta.GetOptions{})
+	if err != nil {
+		return "", errors.WithMessagef(err, "Failed to get credentials from %s secret", secretName)
+	}
+
+	bytes, ok := secret.Data["kubeconfig"]
```

Comment:
I suggest changing the name in the secret to something like `credentials` as this is not a `kubeconfig`.
Also it would be good to extract it to constant.